File: CSharp/ByRound/2137/2137_F_Prefix_Maximum_Invariance.cs
// Problem: CF 2137 F - Prefix Maximum Invariance
// https://codeforces.com/contest/2137/problem/F

/*
F. Prefix Maximum Invariance

Algorithms/Techniques:
- Fenwick Tree (Binary Indexed Tree) for efficient range maximum queries and updates
- Offline processing with coordinate compression (reverse mapping)
- Sliding window technique combined with prefix logic to compute contributions

Time Complexity: O(n log n) per test case due to Fenwick tree operations and sorting
Space Complexity: O(n) for storing the input arrays and the Fenwick tree

Problem Summary:
Given two arrays a and b of size n, we want to compute the sum of f(al, ..., ar, bl, ..., br)
over all subarrays [l,r]. Here f(x,y) is the maximum number of positions where z_i = y_i
such that prefix maxima of z match those of x.

The approach processes each element from left to right using a Fenwick tree indexed by reversed values.
It maintains information about how many elements have been seen so far, enabling efficient queries on 
prefix maximums and counting overlapping elements under constraint of maintaining prefix maxima invariants.
*/

using System;
using System.Collections.Generic;
using System.Linq;
class FenwickMax {
  private int size;
  private int[] tree;
  public void Init(int n) {
    size = n;
    tree = new int[n + 1];
  }
  public void Update(int index, int value) {
    // Update the Fenwick Tree at index with value using bit manipulation
    for (; index <= size; index += index & -index) {
      tree[index] = Math.Max(tree[index], value);
    }
  }
  public int Query(int index) {
    // Query maximum value up to index using bit manipulation
    int result = 0;
    for (; index > 0; index -= index & -index) {
      result = Math.Max(result, tree[index]);
    }
    return result;
  }
}
class Program {
  static void Main() {
    string? line = Console.ReadLine();
    if (line == null)
      return;
    int t = int.Parse(line);
    while (t-- > 0) {
      int n = int.Parse(Console.ReadLine()!);
      int[] a =
          Console.ReadLine()!.Split().Select(int.Parse).Prepend(0).ToArray(); // prepend 0 for 1-based indexing
      int[] b =
          Console.ReadLine()!.Split().Select(int.Parse).Prepend(0).ToArray(); // prepend 0 for 1-based indexing
      int maxVal = 2 * n;
      int Reverse(int v) => maxVal - v + 1; // coordinate compression to avoid negative indices
      var bit = new FenwickMax();
      bit.Init(maxVal + 2);
      long sum1 = 0, sum2 = 0;
      for (int i = 1; i <= n; i++) {
        // Find how many previous elements are >= a[i]: used to know where to place current element in z
        int prevGEa = bit.Query(Reverse(a[i]));
        // Update the Fenwick tree with current index of a[i]
        bit.Update(Reverse(a[i]), i);
        // Find how many previous elements are >= b[i]:
        int LGEb = bit.Query(Reverse(b[i]));
        long weight = n - i + 1; // number of subarrays starting at this position and ending in [i, n]
        sum1 += weight * LGEb; // contribution to sum1 from valid placements
        if (b[i] < a[i]) {
          // If b[i] < a[i], then the element b[i] can be chosen only after some previous element >= a[i]
          sum2 += weight * (i - prevGEa); // penalty term for overlapping elements
        }
      }
      Console.WriteLine(sum1 - sum2);
    }
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByRound/2137/2137_F_Prefix_Maximum_Invariance.py
# Problem: CF 2137 F - Prefix Maximum Invariance
# https://codeforces.com/contest/2137/problem/F

"""
F. Prefix Maximum Invariance
Algorithm: Monotonic stack with binary search optimization
Time Complexity: O(n log n) per test case
Space Complexity: O(n)

The problem involves computing the sum of maximum number of positions where 
z[i] = y[i] over all possible arrays z that have the same prefix maxima as x.
We use a monotonic stack to efficiently track valid candidates for optimal z.

For each subarray [l, r], we compute f(a[l:r+1], b[l:r+1]) which equals:
- For each position i in the subarray, we try to place b[i] in z[i]
- We want z[i] such that prefix maxima stay the same as in a
- The answer is the count of positions where this can be done optimally

The key insight is that for a valid array z, at each position i,
z[i] ≤ max(a[0:i]) (to maintain prefix maxima).
We use stack + binary search to efficiently compute how many such placements work.
"""

import sys

input = lambda: sys.stdin.readline().rstrip()


def solve646():
    n = int(input())
    a = list(map(int, input().split()))
    b = list(map(int, input().split()))
    s = []  # monotonic stack storing indices
    res = 0
    for i in range(n):
        # Pop elements from stack while current element is greater than stack top's value
        while s and a[i] > a[s[-1]]:
            s.pop()
        
        t = 0
        if a[i] ^ b[i]:  # XOR check: if a[i] != b[i]
            # If stack has elements and the max in stack is >= b[i], we find correct position using binary search
            if s and a[s[0]] >= b[i]:
                l, r = 0, len(s)
                while l + 1 < r:
                    m = l + r >> 1
                    if a[s[m]] >= b[i]:
                        l = m
                    else:
                        r = m
                t = (n - i) * (s[l] + 1)
        else:
            # If a[i] == b[i], all positions from s[0] to i could be valid with full overlap
            t = (i + 1) * (n - i)
        res += t
        s.append(i)
    return res


res_test = []
for _ in range(int(input())):
    res_test.append(solve646())
sys.stdout.write("\n".join(map(str, res_test)))


# https://github.com/VaHiX/codeForces/